src/chain_spec.rs

```diff
@@ -139,202 +160,305 @@ fn currency_codes() -> Vec<Ticker> {
+    let mut identity_counter = num_initial_identities;
+    let authority_identities = initial_authorities
+        .iter()
+        .map(|x| {
+            identity_counter += 1;
+            let did = IdentityId::from(identity_counter);
+            let investor_uid = InvestorUid::from(did.as_ref());
+            (x.1.clone(), IdentityId::from(1), did, investor_uid, None)
+        })
+        .collect::<Vec<_>>();
+
+    let all_identities = initial_identities
+        .iter()
+        .cloned()
+        .chain(authority_identities.iter().cloned())
+        .collect::<Vec<_>>();
+    identity_counter = num_initial_identities;
+
+    let secondary_keys = initial_authorities
+        .iter()
+        .map(|x| {
+            identity_counter += 1;
+            (x.0.clone(), IdentityId::from(identity_counter))
+        })
+        .collect::<Vec<_>>();
+
+    let stakers = authority_identities
+        .iter()
+        .cloned()
+        .zip(initial_authorities.iter().cloned())
+        .map(|((_, _, did, ..), x)| {
+            (
+                did,
+                x.0.clone(),
+                x.1.clone(),
+                STASH,
+                general::StakerStatus::Validator,
+            )
+        })
+        .collect::<Vec<_>>();
+
+    (stakers, all_identities, secondary_keys)
+}
+
+fn balances(inits: &[InitialAuth], endoweds: &[AccountId]) -> Vec<(AccountId, u128)> {
+    endoweds
+        .iter()
+        .map(|k: &AccountId| (k.clone(), ENDOWMENT))
+        .chain(inits.iter().map(|x| (x.1.clone(), ENDOWMENT)))
+        .chain(inits.iter().map(|x| (x.0.clone(), STASH)))
+        .collect()
+}
+
+fn bridge_signers() -> Vec<Signatory<AccountId>> {
+    let signer =
+        |seed| Signatory::Account(AccountId::from(get_from_seed::<sr25519::Public>(seed).0));
+    vec![
+        signer("relay_1"),
+        signer("relay_2"),
+        signer("relay_3"),
+        signer("relay_4"),
+        signer("relay_5"),
+    ]
+}
+
+macro_rules! session {
+    ($module:ident, $inits:expr, $build:expr) => {
+        $module::SessionConfig {
+            keys: $inits
+                .iter()
+                .map(|x| {
+                    let sks = $build(x.2.clone(), x.3.clone(), x.4.clone(), x.5.clone());
+                    (x.0.clone(), x.0.clone(), sks)
+                })
+                .collect::<Vec<_>>(),
+        }
+    };
+}
+
+macro_rules! staking {
+    ($auths:expr, $stakers:expr, $cap:expr) => {
+        pallet_staking::GenesisConfig {
+            minimum_validator_count: 1,
+            validator_count: $auths.len() as u32,
+            validator_commission_cap: $cap,
+            stakers: $stakers,
+            invulnerables: $auths.iter().map(|x| x.0.clone()).collect(),
+            slash_reward_fraction: general::Perbill::from_percent(10),
+            min_bond_threshold: 5_000_000_000_000,
+            ..Default::default()
+        }
+    };
+}
+
+macro_rules! im_online {
```

Comment:
IMO it will be cleaner to use functions for these definitions instead of macros.

Reply:
However, it is sadly not possible without getting into deep generics-soup, I tried. ;)